// @budibase/types/dist/documents/app/automation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutomationStatus = exports.AutomationActionStepId = exports.AutomationTriggerStepId = void 0;
var AutomationTriggerStepId;
(function (AutomationTriggerStepId) {
    AutomationTriggerStepId["ROW_SAVED"] = "ROW_SAVED";
    AutomationTriggerStepId["ROW_UPDATED"] = "ROW_UPDATED";
    AutomationTriggerStepId["ROW_DELETED"] = "ROW_DELETED";
    AutomationTriggerStepId["WEBHOOK"] = "WEBHOOK";
    AutomationTriggerStepId["APP"] = "APP";
    AutomationTriggerStepId["CRON"] = "CRON";
})(AutomationTriggerStepId = exports.AutomationTriggerStepId || (exports.AutomationTriggerStepId = {}));
var AutomationActionStepId;
(function (AutomationActionStepId) {
    AutomationActionStepId["SEND_EMAIL_SMTP"] = "SEND_EMAIL_SMTP";
    AutomationActionStepId["CREATE_ROW"] = "CREATE_ROW";
    AutomationActionStepId["UPDATE_ROW"] = "UPDATE_ROW";
    AutomationActionStepId["DELETE_ROW"] = "DELETE_ROW";
    AutomationActionStepId["EXECUTE_BASH"] = "EXECUTE_BASH";
    AutomationActionStepId["OUTGOING_WEBHOOK"] = "OUTGOING_WEBHOOK";
    AutomationActionStepId["EXECUTE_SCRIPT"] = "EXECUTE_SCRIPT";
    AutomationActionStepId["EXECUTE_QUERY"] = "EXECUTE_QUERY";
    AutomationActionStepId["SERVER_LOG"] = "SERVER_LOG";
    AutomationActionStepId["DELAY"] = "DELAY";
    AutomationActionStepId["FILTER"] = "FILTER";
    AutomationActionStepId["QUERY_ROWS"] = "QUERY_ROWS";
    AutomationActionStepId["LOOP"] = "LOOP";
    // these used to be lowercase step IDs, maintain for backwards compat
    AutomationActionStepId["discord"] = "discord";
    AutomationActionStepId["slack"] = "slack";
    AutomationActionStepId["zapier"] = "zapier";
    AutomationActionStepId["integromat"] = "integromat";
})(AutomationActionStepId = exports.AutomationActionStepId || (exports.AutomationActionStepId = {}));
var AutomationStatus;
(function (AutomationStatus) {
    AutomationStatus["SUCCESS"] = "success";
    AutomationStatus["ERROR"] = "error";
    AutomationStatus["STOPPED"] = "stopped";
    AutomationStatus["STOPPED_ERROR"] = "stopped_error";
})(AutomationStatus = exports.AutomationStatus || (exports.AutomationStatus = {}));
