// @budibase/backend-core/dist/src/users.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.searchGlobalUsersByEmail = exports.getGlobalUserByAppPage = exports.searchGlobalUsersByApp = exports.getGlobalUserByEmail = exports.bulkUpdateGlobalUsers = exports.bulkGetGlobalUsersById = void 0;
const db_1 = require("./db");
const context_1 = require("./context");
const bulkGetGlobalUsersById = (userIds) => __awaiter(void 0, void 0, void 0, function* () {
    const db = (0, context_1.getGlobalDB)();
    return (yield db.allDocs({
        keys: userIds,
        include_docs: true,
    })).rows.map(row => row.doc);
});
exports.bulkGetGlobalUsersById = bulkGetGlobalUsersById;
const bulkUpdateGlobalUsers = (users) => __awaiter(void 0, void 0, void 0, function* () {
    const db = (0, context_1.getGlobalDB)();
    return (yield db.bulkDocs(users));
});
exports.bulkUpdateGlobalUsers = bulkUpdateGlobalUsers;
/**
 * Given an email address this will use a view to search through
 * all the users to find one with this email address.
 * @param {string} email the email to lookup the user by.
 */
const getGlobalUserByEmail = (email) => __awaiter(void 0, void 0, void 0, function* () {
    if (email == null) {
        throw "Must supply an email address to view";
    }
    const response = yield (0, db_1.queryGlobalView)(db_1.ViewName.USER_BY_EMAIL, {
        key: email.toLowerCase(),
        include_docs: true,
    });
    if (Array.isArray(response)) {
        // shouldn't be able to happen, but need to handle just in case
        throw new Error(`Multiple users found with email address: ${email}`);
    }
    return response;
});
exports.getGlobalUserByEmail = getGlobalUserByEmail;
const searchGlobalUsersByApp = (appId, opts) => __awaiter(void 0, void 0, void 0, function* () {
    if (typeof appId !== "string") {
        throw new Error("Must provide a string based app ID");
    }
    const params = (0, db_1.getUsersByAppParams)(appId, {
        include_docs: true,
    });
    params.startkey = opts && opts.startkey ? opts.startkey : params.startkey;
    let response = yield (0, db_1.queryGlobalView)(db_1.ViewName.USER_BY_APP, params);
    if (!response) {
        response = [];
    }
    return Array.isArray(response) ? response : [response];
});
exports.searchGlobalUsersByApp = searchGlobalUsersByApp;
const getGlobalUserByAppPage = (appId, user) => {
    if (!user) {
        return;
    }
    return (0, db_1.generateAppUserID)((0, db_1.getProdAppID)(appId), user._id);
};
exports.getGlobalUserByAppPage = getGlobalUserByAppPage;
/**
 * Performs a starts with search on the global email view.
 */
const searchGlobalUsersByEmail = (email, opts) => __awaiter(void 0, void 0, void 0, function* () {
    if (typeof email !== "string") {
        throw new Error("Must provide a string to search by");
    }
    const lcEmail = email.toLowerCase();
    // handle if passing up startkey for pagination
    const startkey = opts && opts.startkey ? opts.startkey : lcEmail;
    let response = yield (0, db_1.queryGlobalView)(db_1.ViewName.USER_BY_EMAIL, Object.assign(Object.assign({}, opts), { startkey, endkey: `${lcEmail}${db_1.UNICODE_MAX}` }));
    if (!response) {
        response = [];
    }
    return Array.isArray(response) ? response : [response];
});
exports.searchGlobalUsersByEmail = searchGlobalUsersByEmail;
