// @budibase/backend-core/dist/src/redis/init.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLockClient = exports.getWritethroughClient = exports.getCacheClient = exports.getAppClient = exports.getSessionClient = exports.getUserClient = void 0;
const redis_1 = __importDefault(require("./redis"));
const utils = __importStar(require("./utils"));
let userClient, sessionClient, appClient, cacheClient, writethroughClient, lockClient;
function init() {
    return __awaiter(this, void 0, void 0, function* () {
        userClient = yield new redis_1.default(utils.Databases.USER_CACHE).init();
        sessionClient = yield new redis_1.default(utils.Databases.SESSIONS).init();
        appClient = yield new redis_1.default(utils.Databases.APP_METADATA).init();
        cacheClient = yield new redis_1.default(utils.Databases.GENERIC_CACHE).init();
        lockClient = yield new redis_1.default(utils.Databases.LOCKS).init();
        writethroughClient = yield new redis_1.default(utils.Databases.WRITE_THROUGH, utils.SelectableDatabase.WRITE_THROUGH).init();
    });
}
process.on("exit", () => __awaiter(void 0, void 0, void 0, function* () {
    if (userClient)
        yield userClient.finish();
    if (sessionClient)
        yield sessionClient.finish();
    if (appClient)
        yield appClient.finish();
    if (cacheClient)
        yield cacheClient.finish();
    if (writethroughClient)
        yield writethroughClient.finish();
    if (lockClient)
        yield lockClient.finish();
}));
function getUserClient() {
    return __awaiter(this, void 0, void 0, function* () {
        if (!userClient) {
            yield init();
        }
        return userClient;
    });
}
exports.getUserClient = getUserClient;
function getSessionClient() {
    return __awaiter(this, void 0, void 0, function* () {
        if (!sessionClient) {
            yield init();
        }
        return sessionClient;
    });
}
exports.getSessionClient = getSessionClient;
function getAppClient() {
    return __awaiter(this, void 0, void 0, function* () {
        if (!appClient) {
            yield init();
        }
        return appClient;
    });
}
exports.getAppClient = getAppClient;
function getCacheClient() {
    return __awaiter(this, void 0, void 0, function* () {
        if (!cacheClient) {
            yield init();
        }
        return cacheClient;
    });
}
exports.getCacheClient = getCacheClient;
function getWritethroughClient() {
    return __awaiter(this, void 0, void 0, function* () {
        if (!writethroughClient) {
            yield init();
        }
        return writethroughClient;
    });
}
exports.getWritethroughClient = getWritethroughClient;
function getLockClient() {
    return __awaiter(this, void 0, void 0, function* () {
        if (!lockClient) {
            yield init();
        }
        return lockClient;
    });
}
exports.getLockClient = getLockClient;
