// @budibase/backend-core/dist/src/db/conversions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractAppUUID = exports.getProdAppID = exports.getDevAppID = exports.getDevelopmentAppID = exports.isDevApp = exports.isProdAppID = exports.isDevAppID = void 0;
const constants_1 = require("../constants");
const NO_APP_ERROR = "No app provided";
function isDevAppID(appId) {
    if (!appId) {
        throw NO_APP_ERROR;
    }
    return appId.startsWith(constants_1.APP_DEV_PREFIX);
}
exports.isDevAppID = isDevAppID;
function isProdAppID(appId) {
    if (!appId) {
        throw NO_APP_ERROR;
    }
    return appId.startsWith(constants_1.APP_PREFIX) && !isDevAppID(appId);
}
exports.isProdAppID = isProdAppID;
function isDevApp(app) {
    if (!app) {
        throw NO_APP_ERROR;
    }
    return isDevAppID(app.appId);
}
exports.isDevApp = isDevApp;
/**
 * Generates a development app ID from a real app ID.
 * @returns {string} the dev app ID which can be used for dev database.
 */
function getDevelopmentAppID(appId) {
    if (!appId || appId.startsWith(constants_1.APP_DEV_PREFIX)) {
        return appId;
    }
    // split to take off the app_ element, then join it together incase any other app_ exist
    const split = appId.split(constants_1.APP_PREFIX);
    split.shift();
    const rest = split.join(constants_1.APP_PREFIX);
    return `${constants_1.APP_DEV_PREFIX}${rest}`;
}
exports.getDevelopmentAppID = getDevelopmentAppID;
exports.getDevAppID = getDevelopmentAppID;
/**
 * Convert a development app ID to a deployed app ID.
 */
function getProdAppID(appId) {
    if (!appId || !appId.startsWith(constants_1.APP_DEV_PREFIX)) {
        return appId;
    }
    // split to take off the app_dev element, then join it together incase any other app_ exist
    const split = appId.split(constants_1.APP_DEV_PREFIX);
    split.shift();
    const rest = split.join(constants_1.APP_DEV_PREFIX);
    return `${constants_1.APP_PREFIX}${rest}`;
}
exports.getProdAppID = getProdAppID;
function extractAppUUID(id) {
    const split = (id === null || id === void 0 ? void 0 : id.split("_")) || [];
    return split.length ? split[split.length - 1] : null;
}
exports.extractAppUUID = extractAppUUID;
