// @budibase/backend-core/dist/src/security/sessions.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSession = exports.endSession = exports.updateSessionTTL = exports.createASession = exports.invalidateSessions = exports.getSessionsForUser = void 0;
const redis = require("../redis/init");
const { v4: uuidv4 } = require("uuid");
const { logWarn } = require("../logging");
const env = require("../environment");
// a week in seconds
const EXPIRY_SECONDS = 86400 * 7;
function makeSessionID(userId, sessionId) {
    return `${userId}/${sessionId}`;
}
function getSessionsForUser(userId) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!userId) {
            console.trace("Cannot get sessions for undefined userId");
            return [];
        }
        const client = yield redis.getSessionClient();
        const sessions = yield client.scan(userId);
        return sessions.map(session => session.value);
    });
}
exports.getSessionsForUser = getSessionsForUser;
function invalidateSessions(userId, opts = {}) {
    return __awaiter(this, void 0, void 0, function* () {
        try {
            const reason = (opts === null || opts === void 0 ? void 0 : opts.reason) || "unknown";
            let sessionIds = opts.sessionIds || [];
            let sessionKeys;
            // If no sessionIds, get all the sessions for the user
            if (sessionIds.length === 0) {
                const sessions = yield getSessionsForUser(userId);
                sessionKeys = sessions.map(session => ({
                    key: makeSessionID(session.userId, session.sessionId),
                }));
            }
            else {
                // use the passed array of sessionIds
                sessionIds = Array.isArray(sessionIds) ? sessionIds : [sessionIds];
                sessionKeys = sessionIds.map(sessionId => ({
                    key: makeSessionID(userId, sessionId),
                }));
            }
            if (sessionKeys && sessionKeys.length > 0) {
                const client = yield redis.getSessionClient();
                const promises = [];
                for (let sessionKey of sessionKeys) {
                    promises.push(client.delete(sessionKey.key));
                }
                if (!env.isTest()) {
                    logWarn(`Invalidating sessions for ${userId} (reason: ${reason}) - ${sessionKeys
                        .map(sessionKey => sessionKey.key)
                        .join(", ")}`);
                }
                yield Promise.all(promises);
            }
        }
        catch (err) {
            console.error(`Error invalidating sessions: ${err}`);
        }
    });
}
exports.invalidateSessions = invalidateSessions;
function createASession(userId, createSession) {
    return __awaiter(this, void 0, void 0, function* () {
        // invalidate all other sessions
        yield invalidateSessions(userId, { reason: "creation" });
        const client = yield redis.getSessionClient();
        const sessionId = createSession.sessionId;
        const csrfToken = createSession.csrfToken ? createSession.csrfToken : uuidv4();
        const key = makeSessionID(userId, sessionId);
        const session = Object.assign(Object.assign({}, createSession), { csrfToken, createdAt: new Date().toISOString(), lastAccessedAt: new Date().toISOString(), userId });
        yield client.store(key, session, EXPIRY_SECONDS);
        return session;
    });
}
exports.createASession = createASession;
function updateSessionTTL(session) {
    return __awaiter(this, void 0, void 0, function* () {
        const client = yield redis.getSessionClient();
        const key = makeSessionID(session.userId, session.sessionId);
        session.lastAccessedAt = new Date().toISOString();
        yield client.store(key, session, EXPIRY_SECONDS);
    });
}
exports.updateSessionTTL = updateSessionTTL;
function endSession(userId, sessionId) {
    return __awaiter(this, void 0, void 0, function* () {
        const client = yield redis.getSessionClient();
        yield client.delete(makeSessionID(userId, sessionId));
    });
}
exports.endSession = endSession;
function getSession(userId, sessionId) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!userId || !sessionId) {
            throw new Error(`Invalid session details - ${userId} - ${sessionId}`);
        }
        const client = yield redis.getSessionClient();
        const session = yield client.get(makeSessionID(userId, sessionId));
        if (!session) {
            throw new Error(`Session not found - ${userId} - ${sessionId}`);
        }
        return session;
    });
}
exports.getSession = getSession;
