// @budibase/backend-core/dist/src/middleware/passport/datasource/google.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.postAuth = exports.preAuth = void 0;
const google = __importStar(require("../google"));
const constants_1 = require("../../../constants");
const utils_1 = require("../../../utils");
const db_1 = require("../../../db");
const environment_1 = __importDefault(require("../../../environment"));
const tenancy_1 = require("../../../tenancy");
const GoogleStrategy = require("passport-google-oauth").OAuth2Strategy;
function fetchGoogleCreds() {
    return __awaiter(this, void 0, void 0, function* () {
        // try and get the config from the tenant
        const db = (0, tenancy_1.getGlobalDB)();
        const googleConfig = yield (0, db_1.getScopedConfig)(db, {
            type: constants_1.Config.GOOGLE,
        });
        // or fall back to env variables
        return (googleConfig || {
            clientID: environment_1.default.GOOGLE_CLIENT_ID,
            clientSecret: environment_1.default.GOOGLE_CLIENT_SECRET,
        });
    });
}
function preAuth(passport, ctx, next) {
    return __awaiter(this, void 0, void 0, function* () {
        // get the relevant config
        const googleConfig = yield fetchGoogleCreds();
        const platformUrl = yield (0, db_1.getPlatformUrl)({ tenantAware: false });
        let callbackUrl = `${platformUrl}/api/global/auth/datasource/google/callback`;
        const strategy = yield google.strategyFactory(googleConfig, callbackUrl);
        if (!ctx.query.appId || !ctx.query.datasourceId) {
            ctx.throw(400, "appId and datasourceId query params not present.");
        }
        return passport.authenticate(strategy, {
            scope: ["profile", "email", "https://www.googleapis.com/auth/spreadsheets"],
            accessType: "offline",
            prompt: "consent",
        })(ctx, next);
    });
}
exports.preAuth = preAuth;
function postAuth(passport, ctx, next) {
    return __awaiter(this, void 0, void 0, function* () {
        // get the relevant config
        const config = yield fetchGoogleCreds();
        const platformUrl = yield (0, db_1.getPlatformUrl)({ tenantAware: false });
        let callbackUrl = `${platformUrl}/api/global/auth/datasource/google/callback`;
        const authStateCookie = (0, utils_1.getCookie)(ctx, constants_1.Cookie.DatasourceAuth);
        return passport.authenticate(new GoogleStrategy({
            clientID: config.clientID,
            clientSecret: config.clientSecret,
            callbackURL: callbackUrl,
        }, (accessToken, refreshToken, profile, done) => {
            (0, utils_1.clearCookie)(ctx, constants_1.Cookie.DatasourceAuth);
            done(null, { accessToken, refreshToken });
        }), { successRedirect: "/", failureRedirect: "/error" }, (err, tokens) => __awaiter(this, void 0, void 0, function* () {
            // update the DB for the datasource with all the user info
            yield (0, db_1.doWithDB)(authStateCookie.appId, (db) => __awaiter(this, void 0, void 0, function* () {
                const datasource = yield db.get(authStateCookie.datasourceId);
                if (!datasource.config) {
                    datasource.config = {};
                }
                datasource.config.auth = Object.assign({ type: "google" }, tokens);
                yield db.put(datasource);
                ctx.redirect(`/builder/app/${authStateCookie.appId}/data/datasource/${authStateCookie.datasourceId}`);
            }));
        }))(ctx, next);
    });
}
exports.postAuth = postAuth;
