// @budibase/backend-core/dist/src/objectStore/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.budibaseTempDir = exports.ObjectStoreBuckets = void 0;
const path_1 = require("path");
const os_1 = require("os");
const fs_1 = __importDefault(require("fs"));
const environment_1 = __importDefault(require("../environment"));
/****************************************************
 *      NOTE: When adding a new bucket - name       *
 *     sure that S3 usages (like budibase-infra)    *
 *  have been updated to have a unique bucket name. *
 ****************************************************/
// can't be an enum - only numbers can be used for computed types
exports.ObjectStoreBuckets = {
    BACKUPS: environment_1.default.BACKUPS_BUCKET_NAME,
    APPS: environment_1.default.APPS_BUCKET_NAME,
    TEMPLATES: environment_1.default.TEMPLATES_BUCKET_NAME,
    GLOBAL: environment_1.default.GLOBAL_BUCKET_NAME,
    GLOBAL_CLOUD: environment_1.default.GLOBAL_CLOUD_BUCKET_NAME,
    PLUGINS: environment_1.default.PLUGIN_BUCKET_NAME,
};
const bbTmp = (0, path_1.join)((0, os_1.tmpdir)(), ".budibase");
if (!fs_1.default.existsSync(bbTmp)) {
    fs_1.default.mkdirSync(bbTmp);
}
function budibaseTempDir() {
    return bbTmp;
}
exports.budibaseTempDir = budibaseTempDir;
