// @budibase/backend-core/dist/src/middleware/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const jwt = __importStar(require("./passport/jwt"));
const local = __importStar(require("./passport/local"));
const google = __importStar(require("./passport/google"));
const oidc = __importStar(require("./passport/oidc"));
const utils_1 = require("./passport/utils");
const authenticated_1 = __importDefault(require("./authenticated"));
const auditLog_1 = __importDefault(require("./auditLog"));
const tenancy_1 = __importDefault(require("./tenancy"));
const internalApi_1 = __importDefault(require("./internalApi"));
const datasourceGoogle = __importStar(require("./passport/datasource/google"));
const csrf_1 = __importDefault(require("./csrf"));
const adminOnly_1 = __importDefault(require("./adminOnly"));
const builderOrAdmin_1 = __importDefault(require("./builderOrAdmin"));
const builderOnly_1 = __importDefault(require("./builderOnly"));
const joiValidator = __importStar(require("./joi-validator"));
const pkg = {
    google,
    oidc,
    jwt,
    local,
    authenticated: authenticated_1.default,
    auditLog: auditLog_1.default,
    tenancy: tenancy_1.default,
    authError: utils_1.authError,
    internalApi: internalApi_1.default,
    ssoCallbackUrl: utils_1.ssoCallbackUrl,
    datasource: {
        google: datasourceGoogle,
    },
    csrf: csrf_1.default,
    adminOnly: adminOnly_1.default,
    builderOnly: builderOnly_1.default,
    builderOrAdmin: builderOrAdmin_1.default,
    joiValidator,
};
module.exports = pkg;
