// @budibase/backend-core/dist/src/errors/licensing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeatureDisabledError = exports.UsageLimitError = exports.context = exports.codes = exports.type = void 0;
const http_1 = require("./http");
exports.type = "license_error";
exports.codes = {
    USAGE_LIMIT_EXCEEDED: "usage_limit_exceeded",
    FEATURE_DISABLED: "feature_disabled",
};
exports.context = {
    [exports.codes.USAGE_LIMIT_EXCEEDED]: (err) => {
        return {
            limitName: err.limitName,
        };
    },
    [exports.codes.FEATURE_DISABLED]: (err) => {
        return {
            featureName: err.featureName,
        };
    },
};
class UsageLimitError extends http_1.HTTPError {
    constructor(message, limitName) {
        super(message, 400, exports.codes.USAGE_LIMIT_EXCEEDED, exports.type);
        this.limitName = limitName;
    }
}
exports.UsageLimitError = UsageLimitError;
class FeatureDisabledError extends http_1.HTTPError {
    constructor(message, featureName) {
        super(message, 400, exports.codes.FEATURE_DISABLED, exports.type);
        this.featureName = featureName;
    }
}
exports.FeatureDisabledError = FeatureDisabledError;
