// @budibase/backend-core/dist/src/cache/generic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bustCache = exports.withCache = exports.destroy = exports.store = exports.get = exports.keys = exports.TTL = exports.CacheKey = void 0;
const BaseCache = require("./base");
const GENERIC = new BaseCache();
var CacheKey;
(function (CacheKey) {
    CacheKey["CHECKLIST"] = "checklist";
    CacheKey["INSTALLATION"] = "installation";
    CacheKey["ANALYTICS_ENABLED"] = "analyticsEnabled";
    CacheKey["UNIQUE_TENANT_ID"] = "uniqueTenantId";
    CacheKey["EVENTS"] = "events";
    CacheKey["BACKFILL_METADATA"] = "backfillMetadata";
    CacheKey["EVENTS_RATE_LIMIT"] = "eventsRateLimit";
})(CacheKey = exports.CacheKey || (exports.CacheKey = {}));
var TTL;
(function (TTL) {
    TTL[TTL["ONE_MINUTE"] = 600] = "ONE_MINUTE";
    TTL[TTL["ONE_HOUR"] = 3600] = "ONE_HOUR";
    TTL[TTL["ONE_DAY"] = 86400] = "ONE_DAY";
})(TTL = exports.TTL || (exports.TTL = {}));
function performExport(funcName) {
    return (...args) => GENERIC[funcName](...args);
}
exports.keys = performExport("keys");
exports.get = performExport("get");
exports.store = performExport("store");
exports.destroy = performExport("delete");
exports.withCache = performExport("withCache");
exports.bustCache = performExport("bustCache");
